File: src/java/hangman.java
import java.util.Scanner;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import java.util.Random;
import java.util.List;
import java.util.ArrayList;
import java.util.HashSet;

import java.lang.Character;

public class Hangman {
    public static void main(String[] args) {
        System.out.println("=== HANGMAN ===");

        if (args.length < 2) {
            System.err.println("Please provide a file path.");
            System.exit(1);
            return;
        }
        String filename = args[1];

        List<String> words;
        try {
            Path file = new File(filename).toPath();
            words = Files.readAllLines(file);
        } catch (IOException err) {
            System.err.println("Failed to read file");
            System.exit(2);
            return;
        }

        Random randGen = new Random();
        Scanner scanner = new Scanner(System.in);

        System.out.println("\n\n\n\n");

        while (true) {
            int index = randGen.nextInt(words.size());
            String word = words.get(index);
            HashSet<Character> correct = new HashSet<>();
            HashSet<Character> incorrect = new HashSet<>();

            System.out.println("Secret: " + word);
            correct.add('a');
            correct.add('e');

            while (true) {
                for (int i = 0; i < 5; i++) {
                    System.out.print("\033[A\033[K");
                }

                String visible = "";
                boolean is_win = true;
                for (int i = 0; i < word.length(); i++) {
                    char ch = word.charAt(i);
                    if (correct.contains(ch)) {
                        visible += ch;
                    } else {
                        visible += '_';
                        is_win = false;
                    }
                }

                if (is_win) {
                    System.out.println("---------");
                    System.out.println("You win! :)");
                    System.out.printf("The word was: '%s'\n", word);
                    System.out.print("---------");
                    scanner.nextLine();
                    break;
                }
                if (incorrect.size() >= 6) {
                    System.out.println("---------");
                    System.out.println("You lose! :(");
                    System.out.printf("The word was: '%s'\n", word);
                    System.out.print("---------");
                    scanner.nextLine();
                    break;
                }

                System.out.println(visible);
                System.out.printf("Chances: %d\n", 6 - incorrect.size());
                System.out.print("Correct: ");
                printSet(correct);
                System.out.print("Inorrect: ");
                printSet(incorrect);
                System.out.print("Guess: ");

                String line = scanner.nextLine();

                if (line.length() < 1) {
                    continue;
                }
                char guess = line.charAt(0);

                if (word.indexOf(guess) >= 0) {
                    correct.add(guess);
                } else {
                    incorrect.add(guess);
                }
            }
        }
    }

    static void printSet(HashSet<Character> set) {
        int i = 0;
        for (Character ch : set) {
            if (i > 0) {
                System.out.print(", ");
            }
            System.out.print(ch);
            i++;
        }
        System.out.println();
    }
}

File: src/hangman.rs
use std::{ collections::HashSet, env, fs, io::{self, Write},
    time::{SystemTime, UNIX_EPOCH},
};

fn main() -> Result<(), &'static str> {
    println!("=== HANGMAN ===");

    let mut args = env::args().skip(1);
    let Some(filename) = args.next() else {
        return Err("Please provide a file path.");
    };

    let Ok(file) = fs::read_to_string(filename) else {
        return Err("Failed to read file.");
    };
    let words: Vec<_> = file.lines().collect();

    println!("\n\n\n\n\n");

    loop {
        let index = random_int(words.len());
        let word = words[index];

        let mut correct = HashSet::<char>::new();
        let mut incorrect = HashSet::<char>::new();

        loop {
            for _ in 0..5 {
                print!("\x1b[A\x1b[K");
            }

            let mut visible = String::with_capacity(word.len());
            let mut is_win = true;
            for ch in word.chars() {
                if correct.contains(&ch) {
                    visible.push(ch);
                } else {
                    visible.push('_');
                    is_win = false;
                }
            }

            if is_win {
                println!("---------");
                println!("You win! :)");
                println!("The word was: '{}'", word);
                println!("---------");
                read_char();
                break;
            }
            if incorrect.len() >= 6 {
                println!("---------");
                println!("You lose! :(");
                println!("The word was: '{}'", word);
                println!("---------");
                read_char();
                break;
            }

            println!("{}", visible);
            println!("Chances: {}", 6 - incorrect.len());
            print!("Correct: ");
            print_set(&correct);
            print!("Incorrect: ");
            print_set(&incorrect);
            print!("Guess: ");

            let Some(guess) = read_char() else {
                continue;
            };

            if word.contains(guess) {
                correct.insert(guess);
            } else {
                incorrect.insert(guess);
            }
        }
    }
}

fn random_int(max: usize) -> usize {
    let now = SystemTime::now();
    let duration = now.duration_since(UNIX_EPOCH).unwrap();
    let nanos = duration.subsec_nanos() as usize;
    nanos % max
}

fn print_set(set: &HashSet<char>) {
    for (i, ch) in set.iter().enumerate() {
        if i > 0 {
            print!(", ");
        }
        print!("{}", ch);
    }
    println!();
}

fn read_char() -> Option<char> {
    let mut string = String::new();
    let _ = io::stdout().flush();
    io::stdin().read_line(&mut string).unwrap();
    string
        .chars()
        .next()
        .filter(|ch| *ch != '\r' && *ch != '\n')
}

